src/Services/BuddyService.cs

```diff
@@ -0,0 +1,179 @@
+
+            return new BuddyActionResult
+            {
+                BuddyUserID = receiver.Uid.ToString(),
+                Result = BuddyActionResultType.Success,
+                Status = relation.Status
+            };
+        }
+
+        public bool RemoveBuddy(Viking owner, Viking receiver)
+        {
+            // get buddy relation 1
+            Model.Buddy relation = ctx.Buddies.Where(e => e.OwnerID == owner.Id)
+                .FirstOrDefault(e => e.BuddyID == receiver.Id);
+
+            // remove it
+            ctx.Buddies.Remove(relation);
+
+            // get buddy relation 2
+
+            Model.Buddy relation2 = ctx.Buddies.Where(e => e.OwnerID == receiver.Id)
+                .FirstOrDefault(e => e.BuddyID == owner.Id);
+
+            // remove it
+            ctx.Buddies.Remove(relation2);
+
+            ctx.SaveChanges();
+
+            return true;
+        }
+
+        public bool AcceptBuddyRequest(Viking owner, Viking receiver)
+        {
+            // get buddy relation 1
+            Model.Buddy relation = ctx.Buddies.Where(e => e.OwnerID == owner.Id)
+                .FirstOrDefault(e => e.BuddyID == receiver.Id);
+
+            // change status to approved
+            relation.Status = BuddyStatus.Approved;
+
+            // get buddy relation 2
+
+            Model.Buddy relation2 = ctx.Buddies.Where(e => e.OwnerID == receiver.Id)
+                .FirstOrDefault(e => e.BuddyID == owner.Id);
+
+            // change status to approved
+            relation2.Status = BuddyStatus.Approved;
+
+            ctx.SaveChanges();
+
+            return true;
+        }
+
+        public BuddyActionResult SetBuddyAsBest(Viking viking, Viking vikingToSet, bool best)
+        {
+            Model.Buddy buddyToSet = ctx.Buddies.Where(e => e.OwnerID == viking.Id)
+                .FirstOrDefault(e => e.BuddyID == vikingToSet.Id);
+
+            if (buddyToSet == null) return new BuddyActionResult { Result = BuddyActionResultType.Unknown };
+
+            buddyToSet.BestBuddy = best;
+            ctx.SaveChanges();
+
+            return new BuddyActionResult { Result = BuddyActionResultType.Success };
+        }
+
+        public BuddyLocation GetBuddyLocation(Viking viking)
+        {
+            Model.Buddy buddy = ctx.Buddies.FirstOrDefault(e => e.BuddyID == viking.Id);
+
+            if (viking.CurrentRoom != null) return new BuddyLocation
+            {
+                Server = config.Value.MMOAdress,
+                ServerPort = config.Value.MMOPort,
+                ServerVersion = "S2X", // always use SmartFox 2X Protocol
+                Zone = "JumpStart", // TODO - put this in config
+                Room = viking.CurrentRoom,
+                UserID = buddy.BuddyID.ToString(),
+                MultiplayerID = 1, // TODO - what the fuck is this
+                AppName = "JSMain" // TODO - figure out how to detect this
```

Comment:
probably this can be obtained from `gameVersion` ... the question is whether the client requires receiving the correct value here?

Reply:
GetBuddyLocation isn't working right now anyways, could comment function and endpoint out for now.